ee/hogai/trends/toolkit.py

```diff
@@ -0,0 +1,498 @@
+from posthog.hogql_queries.query_runner import ExecutionMode
+from posthog.models.group_type_mapping import GroupTypeMapping
+from posthog.models.property_definition import PropertyDefinition, PropertyType
+from posthog.models.team.team import Team
+from posthog.schema import (
+    ActorsPropertyTaxonomyQuery,
+    CachedActorsPropertyTaxonomyQueryResponse,
+    CachedEventTaxonomyQueryResponse,
+    EventTaxonomyQuery,
+    ExperimentalAITrendsQuery,
+)
+
+
+class ToolkitTool(TypedDict):
+    name: str
+    signature: str
+    description: str
+
+
+class RetrieveEntityPropertiesValuesArgsModel(BaseModel):
+    entity: str
+    property_name: str
+
+
+class RetrieveEntityPropertiesValuesModel(BaseModel):
+    name: Literal["retrieve_entity_property_values"]
+    arguments: RetrieveEntityPropertiesValuesArgsModel
+
+
+class RetrieveEventPropertiesValuesArgsModel(BaseModel):
+    event_name: str
+    property_name: str
+
+
+class RetrieveEventPropertiesValuesModel(BaseModel):
+    name: Literal["retrieve_event_property_values"]
+    arguments: RetrieveEventPropertiesValuesArgsModel
+
+
+class SingleArgumentTrendsAgentToolModel(BaseModel):
+    name: Literal[
+        "retrieve_entity_properties",
+        "retrieve_event_properties",
+        "final_answer",
+        "handle_incorrect_response",
+    ]
+    arguments: str
+
+
+class TrendsAgentToolModel(
+    RootModel[
+        Union[
+            SingleArgumentTrendsAgentToolModel, RetrieveEntityPropertiesValuesModel, RetrieveEventPropertiesValuesModel
+        ]
+    ]
+):
+    root: Union[
+        SingleArgumentTrendsAgentToolModel, RetrieveEntityPropertiesValuesModel, RetrieveEventPropertiesValuesModel
+    ] = Field(..., discriminator="name")
+
+
+class TrendsAgentToolkit:
+    _team: Team
+
+    def __init__(self, team: Team):
+        self._team = team
+
+    @property
+    def groups(self):
+        return GroupTypeMapping.objects.filter(team=self._team).order_by("group_type_index")
+
+    @cached_property
+    def _entity_names(self) -> list[str]:
+        """
+        The schemas use `group_type_index` for groups complicating things for the agent. Instead, we use groups' names,
+        so the generation step will handle their indexes. Tools would need to support multiple arguments, or we would need
+        to create various tools for different group types. Since we don't use function calling here, we want to limit the
+        number of tools because non-function calling models can't handle many tools.
+        """
+        entities = [
+            "person",
+            "session",
+            # "cohort", # not supported yet
+            *[group.group_type for group in self.groups],
+        ]
+        return entities
+
+    @cached_property
+    def tools(self) -> list[ToolkitTool]:
+        """
+        Our ReAct agent doesn't use function calling. Instead, it uses tools in natural language to decide next steps. The agent expects the following format:
```

Comment:
_Should_ it be using function calling?

Reply:
I don't think so, at least not for the current state of models. There is an [experiment](https://github.com/PostHog/ai_notebooks/blob/main/trends_function_calling.ipynb) that uses function calling for taxonomy discovery, but it doesn't perform as well as ReAct.

One thing that we may explore is combining a ReAct agent and function calling to simplify output parsing while keeping similar <s>reasoning</s> predicting capabilities. Models without function calling aren't fine-tuned for a structured output, so accuracy is lower for a JSON/XML/whatever structured format.